rust_src/src/base64.rs

```diff
@@ -0,0 +1,127 @@
+extern crate libc;
+extern crate base64;
+
+use base64::base64::{decode, encode};
+use std::ffi::CStr;
+use std::ffi::CString;
+use std::ptr;
+
+const MIME_LINE_LENGTH: usize = 76;
+
+#[no_mangle]
+pub extern "C" fn base64_encode_1(
+    from: *const libc::c_char,
+    to: *mut libc::c_char,
+    length: libc::ptrdiff_t,
+    line_break: bool,
+    multibyte: bool
+)
+    -> libc::ptrdiff_t
+{
+    let bytes = unsafe { CStr::from_ptr(from).to_bytes() };
+    let coded = encode(bytes);
+
+    let string = CString::new(coded).unwrap();
+    let size = string.as_bytes().len() as isize;
+
+    let n = if size < length { size } else { length  };
+
+    if line_break && MIME_LINE_LENGTH < n as usize {
+        let bytes = string.into_bytes();
+
+        let mut sink = to;
+        for chunk in bytes.chunks(MIME_LINE_LENGTH) {
+            let size = chunk.len();
+            let c = chunk.as_ptr();
+
+            unsafe {
+                ptr::copy(c as *const libc::c_char, sink, size);
+                sink = sink.offset(size as isize);
+                *sink = b'\n' as libc::c_char;
+                sink = sink.offset(1);
+            }
+        }
+
+        n
+    } else {
+        unsafe {
+            ptr::copy(string.into_raw(), to, n as usize);
+        }
+        n
+    }
+}
+
+//  Base64-decode the data at FROM of LENGTH bytes into TO.  If
+//  MULTIBYTE, the decoded result should be in multibyte
+//  form.  If NCHARS_RETURN is not NULL, store the number of produced
+//  characters in *NCHARS_RETURN.
+
+#[no_mangle]
+pub extern "C" fn base64_decode_1(
+    from: *const libc::c_char,
+    to: *mut libc::c_char,
+    length: libc::ptrdiff_t,
+    multibyte: bool,
+    nchars_return: *mut libc::ptrdiff_t
+)
+    -> libc::ptrdiff_t
+{
+    match unsafe { CStr::from_ptr(from).to_str() } {
+        Ok(encoded) => {
+            let decoded = CString::new(decode(encoded).unwrap()).unwrap();
```

Comment:
I think it would be good to rework the code to avoid these raw unwraps. It seems like a good end-state would be that if either decode or CString::new, this function should return -1. 

I'm picturing a function which does the decoding, using a rust interface. That way it's easy for other rust code to use this functionality directly, without having to worry about unsafe mutable buffers. Something like

```
let foo = CStr::from_ptr(from)...
match decode_base64(foo) {
    Ok(encoded) => // copy bytes and return n
    Err(_) => -1
}

fn decode_base64(to_decode: &str) -> Result<CString> { // This can be a regular string one day, when we port
                                                       // the remaining base64-encode functions to rust.
       let decoded = try!(decode(to_decode)); // You may need to map_err here. 
       try!(CString::new(decoded)) // or map_err here
}
```

This way we have a clean separation between rust functionality using rust primitives, and c functionality using unsafe c-style code.